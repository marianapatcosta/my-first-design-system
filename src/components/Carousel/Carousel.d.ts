import React, { CSSProperties } from 'react';
import { ORIENTATION } from '../../constants';
declare type orientationTypes = ORIENTATION;
export interface CarouselProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
     * Carousel's title
     */
    title?: string;
    /**
     * array of images to display
     */
    items: Array<{
        src: string;
        name: string;
    }>;
    /**
     * Carousel's orientation
     */
    orientation?: orientationTypes;
    /**
     * is it supposed to change the main image automatically after a certain time interval, i.e without user request?
     */
    allowAutomaticView?: boolean;
    /**
     * is it supposed to show a toggle to alternate between automatic and manual mode?
     */
    canControlAutomaticView?: boolean;
    /**
     * is it supposed to show image previwes beside the main image?
     */
    hasPreviews?: boolean;
    /**
     * time interval to automatically change the main image, in milliseconds, if automatic view is enable
     */
    imageChangeInterval?: number;
}
export declare const Carousel: React.FC<CarouselProps>;
export {};
