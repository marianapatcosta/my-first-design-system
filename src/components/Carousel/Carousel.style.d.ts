import { KeyboardEvent } from 'react';
export declare const StyledCarousel: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledCarouselContent: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledCarouselHeader: import("styled-components").StyledComponent<"header", any, {}, never>;
export declare const StyledCarouselTitle: import("styled-components").StyledComponent<"h2", any, {}, never>;
export declare const StyledCarouselToggle: import("styled-components").StyledComponent<import("react").FC<import("../ToggleSwitch/ToggleSwitch").ToggleSwitchProps>, any, {}, never>;
export declare const StyledChevron: import("styled-components").StyledComponent<"button", any, {
    isLeftChevron?: boolean | undefined;
}, never>;
export declare const StyledCarouselSelected: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledImageWrapper: import("styled-components").StyledComponent<"div", any, {
    isVertical?: boolean | undefined;
}, never>;
export declare const StyledImage: import("styled-components").StyledComponent<import("react").FC<import("../Image/Image").ImageProps>, any, {
    isSelected?: boolean | undefined;
    role?: string | undefined;
}, never>;
export declare const StyledIndicatorBar: import("styled-components").StyledComponent<"span", any, {}, never>;
export declare const StyledIndicatorBarItem: import("styled-components").StyledComponent<"button", any, {
    isFilled?: boolean | undefined;
}, never>;
export declare const StyledPreviews: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledImagePreview: import("styled-components").StyledComponent<import("react").FC<import("../Image/Image").ImageProps>, any, {
    isVertical?: boolean | undefined;
    role?: string | undefined;
    tabIndex?: number | undefined;
    onKeyDown?: ((event: KeyboardEvent<HTMLImageElement>) => void) | undefined;
}, never>;
