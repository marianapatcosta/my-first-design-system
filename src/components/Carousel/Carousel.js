var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useRef, useState, } from 'react';
import { Chevron } from '../../assets/icons';
import { ORIENTATION, PREVIEWS_POSITION } from '../../constants';
import { isEventValid } from '../../utils/utils';
import { StyledCarousel, StyledCarouselContent, StyledCarouselSelected, StyledCarouselHeader, StyledCarouselTitle, StyledCarouselToggle, StyledChevron, StyledIndicatorBar, StyledIndicatorBarItem, StyledImage, StyledImageWrapper, StyledPreviews, StyledImagePreview, } from './Carousel.style';
export var Carousel = function (_a) {
    var title = _a.title, items = _a.items, hasPreviews = _a.hasPreviews, orientation = _a.orientation, imageChangeInterval = _a.imageChangeInterval, allowAutomaticView = _a.allowAutomaticView, canControlAutomaticView = _a.canControlAutomaticView, props = __rest(_a, ["title", "items", "hasPreviews", "orientation", "imageChangeInterval", "allowAutomaticView", "canControlAutomaticView"]);
    var _b = useState(0), presentingItemIndex = _b[0], setPresentingItemIndex = _b[1];
    var _c = useState(allowAutomaticView || false), isAutomaticView = _c[0], setAutomaticView = _c[1];
    var chevronRef = useRef(null);
    useEffect(function () {
        var interval;
        if (isAutomaticView) {
            interval = window.setInterval(function () { var _a; return !!chevronRef && ((_a = chevronRef.current) === null || _a === void 0 ? void 0 : _a.click()); }, imageChangeInterval);
        }
        return function () { return clearInterval(interval); };
    }, [isAutomaticView, imageChangeInterval]);
    var toggleAutomaticView = function (event) {
        return !!event &&
            allowAutomaticView &&
            canControlAutomaticView &&
            setAutomaticView(event.target.checked);
    };
    var handleNextItemClick = function () {
        var newIndex = presentingItemIndex < items.length - 1 ? presentingItemIndex + 1 : 0;
        setPresentingItemIndex(newIndex);
    };
    var handlePreviousItemClick = function () {
        var newIndex = presentingItemIndex > 0 ? presentingItemIndex - 1 : items.length - 1;
        setPresentingItemIndex(newIndex);
    };
    var handleItemClick = function (index) { return setPresentingItemIndex(index); };
    var renderPreviews = function (previewPosition) {
        var nextIndex = presentingItemIndex < items.length - 1 ? presentingItemIndex + 1 : 0;
        var afterNextIndex = nextIndex < items.length - 1 ? nextIndex + 1 : 0;
        var previousIndex = presentingItemIndex > 0 ? presentingItemIndex - 1 : items.length - 1;
        var beforePreviousIndex = previousIndex > 0 ? previousIndex - 1 : items.length - 1;
        var isBefore = previewPosition === PREVIEWS_POSITION.BEFORE;
        return (_jsxs(StyledPreviews, { children: [_jsx(StyledImagePreview, { role: 'button', tabIndex: 0, isVertical: orientation === ORIENTATION.VERTICAL, isClickable: true, src: isBefore ? items[beforePreviousIndex].src : items[nextIndex].src, alt: "preview-" + (isBefore ? items[beforePreviousIndex].name : items[nextIndex].name), onClick: function () {
                        return handleItemClick(isBefore ? beforePreviousIndex : nextIndex);
                    } }, void 0),
                _jsx(StyledImagePreview, { role: 'button', tabIndex: 0, isClickable: true, isVertical: orientation === ORIENTATION.VERTICAL, src: isBefore ? items[previousIndex].src : items[afterNextIndex].src, alt: "preview-" + (isBefore ? items[previousIndex].name : items[afterNextIndex].name), onClick: function () {
                        return handleItemClick(isBefore ? previousIndex : afterNextIndex);
                    } }, void 0)] }, void 0));
    };
    return (_jsx(StyledCarousel, __assign({}, props, { role: 'listbox', "aria-label": 'images slider' }, { children: _jsxs(StyledCarouselContent, { children: [_jsxs(StyledCarouselHeader, { children: [_jsx(StyledCarouselTitle, { children: title }, void 0),
                        allowAutomaticView && canControlAutomaticView && (_jsx(StyledCarouselToggle, { "data-testid": 'carousel-toggle', leftLabel: 'off', rightLabel: 'on', checked: isAutomaticView, onChange: toggleAutomaticView }, void 0))] }, void 0),
                _jsxs(StyledCarouselSelected, { children: [hasPreviews && renderPreviews(PREVIEWS_POSITION.BEFORE), _jsxs(StyledImageWrapper, __assign({ isVertical: orientation === ORIENTATION.VERTICAL }, { children: [items.map(function (item, index) { return (_jsx(StyledImage, { role: 'option', src: item.src, alt: "image-" + item.name, isSelected: presentingItemIndex === index }, "slide-image-" + item.name)); }),
                                _jsx(StyledChevron, __assign({ "aria-label": 'click to select previous item', isLeftChevron: true, onClick: handlePreviousItemClick, onKeyDown: function (event) {
                                        return isEventValid(event) && handlePreviousItemClick();
                                    } }, { children: _jsx("img", { src: Chevron, alt: 'left-chevron' }, void 0) }), void 0),
                                _jsx(StyledChevron, __assign({ "aria-label": 'click to select next item', ref: chevronRef, onClick: handleNextItemClick, onKeyDown: function (event) {
                                        return isEventValid(event) && handleNextItemClick();
                                    } }, { children: _jsx("img", { src: Chevron, alt: 'right-chevron' }, void 0) }), void 0),
                                _jsx(StyledIndicatorBar, { children: items.map(function (item, index) { return (_jsx(StyledIndicatorBarItem, { "data-testid": "bar-item-" + index, "aria-label": "click to select item " + index, isFilled: index <= presentingItemIndex, onClick: function () { return handleItemClick(index); }, onKeyDown: function (event) {
                                            return isEventValid(event) && handleItemClick(index);
                                        } }, "carousel-item-" + index * Math.random())); }) }, void 0)] }), void 0), hasPreviews && renderPreviews(PREVIEWS_POSITION.AFTER)] }, void 0)] }, void 0) }), void 0));
};
Carousel.defaultProps = {
    id: '',
    className: '',
    style: {},
    title: '',
    items: [],
    hasPreviews: false,
    imageChangeInterval: 1500,
    allowAutomaticView: false,
    canControlAutomaticView: false,
    orientation: ORIENTATION.LANDSCAPE,
};
