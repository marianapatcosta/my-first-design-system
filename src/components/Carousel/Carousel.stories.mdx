import { Meta } from '@storybook/addon-docs/blocks';
import { Carousel } from './Carousel';
import { ORIENTATION } from '../../constants';
import Test from '../../assets/images/test.jpg';
import Test1 from '../../assets/images/test1.jpg';
import Test2 from '../../assets/images/test2.jpg';
import Test3 from '../../assets/images/test3.jpg';
import Test4 from '../../assets/images/test4.jpg';
import Test5 from '../../assets/images/test5.jpg';
import Test6 from '../../assets/images/test6.jpg';
import Test7 from '../../assets/images/test7.jpg';

<Meta title='Design System|Carousel' component={Carousel} />

# Carousel

Simple slider component

## Usage

Use Carousel to display multiple images, either on portrait or landscape orientation, but only one of them is to be shown at a time. To change the image, the user
should click on the lateral arrows. This component also allows to activate automatic mode, so the images are
dynamically changed after a customizable time interval. It may also show previews of the two previous and two next images,
if `hasPreviews` prop is true.

- Default Carousel:

  <Story name='Default'>
    <Carousel
      title='Image slider'
      items={[
        { src: Test, name: 'test' },
        { src: Test1, name: 'test1' },
        { src: Test2, name: 'test2' },
        { src: Test3, name: 'test3' },
        { src: Test4, name: 'test4' },
      ]}
    />
  </Story>

- Vertical Carousel:

  <Story name='Vertical'>
    <Carousel
      title='Image slider'
      orientation={ORIENTATION.VERTICAL}
      items={[
        { src: Test5, name: 'test5' },
        { src: Test6, name: 'test6' },
        { src: Test7, name: 'test7' },
      ]}
    />
  </Story>

- Carousel WithPreviews:

  <Story name='WithPreviews'>
    <Carousel
      title='Image slider'
      items={[
        { src: Test, name: 'test' },
        { src: Test1, name: 'test1' },
        { src: Test2, name: 'test2' },
        { src: Test3, name: 'test3' },
        { src: Test4, name: 'test4' },
      ]}
      hasPreviews={true}
    />
  </Story>

- Carousel AutomaticView:

  <Story name='AutomaticView'>
    <Carousel
      title='Image slider'
      items={[
        { src: Test, name: 'test' },
        { src: Test1, name: 'test1' },
        { src: Test2, name: 'test2' },
        { src: Test3, name: 'test3' },
        { src: Test4, name: 'test4' },
      ]}
      allowAutomaticView={true}
    />
  </Story>

  - Carousel Controlled AutomaticView:

  <Story name='ControlledAutomaticView'>
    <Carousel
      title='Image slider'
      items={[
        { src: Test, name: 'test' },
        { src: Test1, name: 'test1' },
        { src: Test2, name: 'test2' },
        { src: Test3, name: 'test3' },
        { src: Test4, name: 'test4' },
      ]}
      allowAutomaticView={true}
      canControlAutomaticView={true}
    />
  </Story>

## Other important notes

not applicable.
