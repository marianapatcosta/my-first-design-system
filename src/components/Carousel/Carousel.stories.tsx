import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Carousel, CarouselProps } from './Carousel';
import Test from '../../assets/images/test.jpg';
import Test1 from '../../assets/images/test1.jpg';
import Test2 from '../../assets/images/test2.jpg';
import Test3 from '../../assets/images/test3.jpg';
import Test4 from '../../assets/images/test4.jpg';
import Test5 from '../../assets/images/test5.jpg';
import Test6 from '../../assets/images/test6.jpg';
import Test7 from '../../assets/images/test7.jpg';
import { ORIENTATION } from '../../constants';

export default {
  title: 'Carousel',
  component: Carousel,
  args: {
    title: 'Image slider',
    items: [
      { src: Test, name: 'test' },
      { src: Test1, name: 'test1' },
      { src: Test2, name: 'test2' },
      { src: Test3, name: 'test3' },
      { src: Test4, name: 'test4' },
    ],
  },
} as Meta;

const Template: Story<CarouselProps> = (args) => <Carousel {...args} />;

export const Default = Template.bind({});
Default.args = {};

export const Vertical = Template.bind({});
Vertical.args = {
  orientation: ORIENTATION.VERTICAL,
  items: [
    { src: Test5, name: 'test5' },
    { src: Test6, name: 'test6' },
    { src: Test7, name: 'test7' },
  ],
};

export const WithPreviews = Template.bind({});
WithPreviews.args = { hasPreviews: true };

export const AutomaticView = Template.bind({});
AutomaticView.args = { allowAutomaticView: true };

export const ControlledAutomaticView = Template.bind({});
ControlledAutomaticView.args = {
  allowAutomaticView: true,
  canControlAutomaticView: true,
};
