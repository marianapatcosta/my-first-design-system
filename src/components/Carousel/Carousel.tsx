import React, {
  ChangeEvent,
  CSSProperties,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Chevron } from '../../assets/icons';
import { ORIENTATION, PREVIEWS_POSITION } from '../../constants';
import { isEventValid } from '../../utils/utils';
import {
  StyledCarousel,
  StyledCarouselContent,
  StyledCarouselSelected,
  StyledCarouselHeader,
  StyledCarouselTitle,
  StyledCarouselToggle,
  StyledChevron,
  StyledIndicatorBar,
  StyledIndicatorBarItem,
  StyledImage,
  StyledImageWrapper,
  StyledPreviews,
  StyledImagePreview,
} from './Carousel.style';

type orientationTypes = ORIENTATION;

export interface CarouselProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
   * Carousel's title
   */
  title?: string;
  /**
   * array of images to display
   */
  items: Array<{ src: string; name: string }>;
  /**
   * Carousel's orientation
   */
  orientation?: orientationTypes;
  /**
   * is it supposed to change the main image automatically after a certain time interval, i.e without user request?
   */
  allowAutomaticView?: boolean;
  /**
   * is it supposed to show a toggle to alternate between automatic and manual mode?
   */
  canControlAutomaticView?: boolean;
  /**
   * is it supposed to show image previwes beside the main image?
   */
  hasPreviews?: boolean;
  /**
   * time interval to automatically change the main image, in milliseconds, if automatic view is enable
   */
  imageChangeInterval?: number;
}

export const Carousel: React.FC<CarouselProps> = ({
  title,
  items,
  hasPreviews,
  orientation,
  imageChangeInterval,
  allowAutomaticView,
  canControlAutomaticView,
  ...props
}) => {
  const [presentingItemIndex, setPresentingItemIndex] = useState<number>(0);
  const [isAutomaticView, setAutomaticView] = useState<boolean>(
    allowAutomaticView || false
  );
  const chevronRef = useRef<HTMLButtonElement>(null);

  useEffect(() => {
    let interval: number;
    if (isAutomaticView) {
      interval = window.setInterval(
        () => !!chevronRef && chevronRef.current?.click(),
        imageChangeInterval
      );
    }
    return () => clearInterval(interval);
  }, [isAutomaticView, imageChangeInterval]);

  const toggleAutomaticView = (event?: ChangeEvent<HTMLInputElement>) =>
    !!event &&
    allowAutomaticView &&
    canControlAutomaticView &&
    setAutomaticView(event.target.checked);

  const handleNextItemClick = () => {
    const newIndex =
      presentingItemIndex < items.length - 1 ? presentingItemIndex + 1 : 0;
    setPresentingItemIndex(newIndex);
  };

  const handlePreviousItemClick = () => {
    const newIndex =
      presentingItemIndex > 0 ? presentingItemIndex - 1 : items.length - 1;
    setPresentingItemIndex(newIndex);
  };

  const handleItemClick = (index: number) => setPresentingItemIndex(index);

  const renderPreviews = (previewPosition: PREVIEWS_POSITION) => {
    const nextIndex: number =
      presentingItemIndex < items.length - 1 ? presentingItemIndex + 1 : 0;
    const afterNextIndex: number =
      nextIndex < items.length - 1 ? nextIndex + 1 : 0;
    const previousIndex: number =
      presentingItemIndex > 0 ? presentingItemIndex - 1 : items.length - 1;
    const beforePreviousIndex: number =
      previousIndex > 0 ? previousIndex - 1 : items.length - 1;
    const isBefore: boolean = previewPosition === PREVIEWS_POSITION.BEFORE;

    return (
      <StyledPreviews>
        <StyledImagePreview
          role='button'
          tabIndex={0}
          isVertical={orientation === ORIENTATION.VERTICAL}
          isClickable
          src={isBefore ? items[beforePreviousIndex].src : items[nextIndex].src}
          alt={`preview-${
            isBefore ? items[beforePreviousIndex].name : items[nextIndex].name
          }`}
          onClick={() =>
            handleItemClick(isBefore ? beforePreviousIndex : nextIndex)
          }
        />
        <StyledImagePreview
          role='button'
          tabIndex={0}
          isClickable
          isVertical={orientation === ORIENTATION.VERTICAL}
          src={isBefore ? items[previousIndex].src : items[afterNextIndex].src}
          alt={`preview-${
            isBefore ? items[previousIndex].name : items[afterNextIndex].name
          }`}
          onClick={() =>
            handleItemClick(isBefore ? previousIndex : afterNextIndex)
          }
        />
      </StyledPreviews>
    );
  };

  return (
    <StyledCarousel {...props} role='listbox' aria-label='images slider'>
      <StyledCarouselContent>
        <StyledCarouselHeader>
          <StyledCarouselTitle>{title}</StyledCarouselTitle>
          {allowAutomaticView && canControlAutomaticView && (
            <StyledCarouselToggle
              data-testid='carousel-toggle'
              leftLabel='off'
              rightLabel='on'
              checked={isAutomaticView}
              onChange={toggleAutomaticView}
            />
          )}
        </StyledCarouselHeader>
        <StyledCarouselSelected>
          {hasPreviews && renderPreviews(PREVIEWS_POSITION.BEFORE)}
          <StyledImageWrapper isVertical={orientation === ORIENTATION.VERTICAL}>
            {items.map((item, index) => (
              <StyledImage
                role='option'
                key={`slide-image-${item.name}`}
                src={item.src}
                alt={`image-${item.name}`}
                isSelected={presentingItemIndex === index}
              />
            ))}
            <StyledChevron
              aria-label='click to select previous item'
              isLeftChevron
              onClick={handlePreviousItemClick}
              onKeyDown={(event) =>
                isEventValid(event) && handlePreviousItemClick()
              }
            >
              <img src={Chevron} alt='left-chevron' />
            </StyledChevron>

            <StyledChevron
              aria-label='click to select next item'
              ref={chevronRef}
              onClick={handleNextItemClick}
              onKeyDown={(event) =>
                isEventValid(event) && handleNextItemClick()
              }
            >
              <img src={Chevron} alt='right-chevron' />
            </StyledChevron>

            <StyledIndicatorBar>
              {items.map((item, index) => (
                <StyledIndicatorBarItem
                  data-testid={`bar-item-${index}`}
                  aria-label={`click to select item ${index}`}
                  key={`carousel-item-${index * Math.random()}`}
                  isFilled={index <= presentingItemIndex}
                  onClick={() => handleItemClick(index)}
                  onKeyDown={(event) =>
                    isEventValid(event) && handleItemClick(index)
                  }
                />
              ))}
            </StyledIndicatorBar>
          </StyledImageWrapper>
          {hasPreviews && renderPreviews(PREVIEWS_POSITION.AFTER)}
        </StyledCarouselSelected>
      </StyledCarouselContent>
    </StyledCarousel>
  );
};

Carousel.defaultProps = {
  id: '',
  className: '',
  style: {},
  title: '',
  items: [],
  hasPreviews: false,
  imageChangeInterval: 1500,
  allowAutomaticView: false,
  canControlAutomaticView: false,
  orientation: ORIENTATION.LANDSCAPE,
};
