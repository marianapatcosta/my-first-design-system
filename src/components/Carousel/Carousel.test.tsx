import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { Carousel } from './Carousel';
import Test from '../../assets/images/test.jpg';
import Test1 from '../../assets/images/test1.jpg';
import Test2 from '../../assets/images/test2.jpg';
import Test3 from '../../assets/images/test3.jpg';
import Test4 from '../../assets/images/test4.jpg';

describe('Carousel', () => {
  const defaultProps = {
    title: 'Image slider',
    items: [
      { src: Test, name: 'test' },
      { src: Test1, name: 'test1' },
      { src: Test2, name: 'test2' },
      { src: Test3, name: 'test3' },
      { src: Test4, name: 'test4' },
    ],
  };
  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<Carousel {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should display the first image item on render', () => {
    const { getByAltText } = render(<Carousel {...defaultProps} />);

    const firstImage = getByAltText('image-test');
    const style = window.getComputedStyle(firstImage);
    expect(style.opacity).toBe('1');
  });

  test('should display the second image item after click on right chevron', () => {
    const { getByAltText } = render(<Carousel {...defaultProps} />);

    const chevron = getByAltText('right-chevron');
    fireEvent.click(chevron);
    const secondImage = getByAltText('image-test1');
    const style = window.getComputedStyle(secondImage);
    expect(style.opacity).toBe('1');
  });

  test('should display the second image item after press `enter` on right chevron', () => {
    const { getByAltText } = render(<Carousel {...defaultProps} />);

    const chevron = getByAltText('right-chevron');
    fireEvent.keyDown(chevron, { key: 'Enter' });
    const secondImage = getByAltText('image-test1');
    const style = window.getComputedStyle(secondImage);
    expect(style.opacity).toBe('1');
  });

  test('should display the last image item after click on left chevron', () => {
    const { getByAltText } = render(<Carousel {...defaultProps} />);

    const chevron = getByAltText('left-chevron');
    fireEvent.keyDown(chevron, { key: 'Enter' });

    const lastImage = getByAltText('image-test4');
    const style = window.getComputedStyle(lastImage);
    expect(style.opacity).toBe('1');
  });

  test('should display the last image item after press `enter` on left chevron', () => {
    const { getByAltText } = render(<Carousel {...defaultProps} />);

    const chevron = getByAltText('left-chevron');
    fireEvent.click(chevron);
    const lastImage = getByAltText('image-test4');
    const style = window.getComputedStyle(lastImage);
    expect(style.opacity).toBe('1');
  });

  test('should display the image previews if hasPreviews prop is passed', () => {
    const { getByAltText } = render(<Carousel {...defaultProps} hasPreviews />);
    const previewThirdImage = getByAltText('preview-test3');
    expect(previewThirdImage).toBeInTheDocument();
  });

  test('should display the image 4 after clicking on image 4 preview', () => {
    const { getByAltText } = render(<Carousel {...defaultProps} hasPreviews />);

    const previewFourthImage = getByAltText('preview-test4');
    fireEvent.click(previewFourthImage);
    const fourthImage = getByAltText('image-test4');
    const style = window.getComputedStyle(fourthImage);
    expect(style.opacity).toBe('1');
  });

  test('should display the image 3 after clicking on image 3 preview', () => {
    const { getByAltText } = render(<Carousel {...defaultProps} hasPreviews />);

    const previewThirdImage = getByAltText('preview-test3');
    fireEvent.click(previewThirdImage);
    const thirdImage = getByAltText('image-test3');
    const style = window.getComputedStyle(thirdImage);
    expect(style.opacity).toBe('1');
  });

  /* test('should automatically change image if `allowAutomaticView` prop is passed', () => {
    const { getByAltText } = render(
      <Carousel {...defaultProps} allowAutomaticView />
    );
    jest.useFakeTimers();
    act(() => {
      jest.runAllTimers();
    });
    jest.useRealTimers();
    const secondImage = getByAltText('image-test1');
    const style = window.getComputedStyle(secondImage);
    expect(style.opacity).toBe('1');
  });

  test('should display a toogle component is `allowAutomaticView` and `canControlAutomaticView` props are passed', () => {
    const { getByTestId } = render(
      <Carousel {...defaultProps} allowAutomaticView canControlAutomaticView />
    );

    const toggle = getByTestId('carousel-toggle');
    expect(toggle).toBeInTheDocument();
  }); */

  test('should display image 3 if the fourth bar item is clicked', () => {
    const { getByTestId, getByAltText } = render(
      <Carousel {...defaultProps} allowAutomaticView />
    );

    const fourthBarItem = getByTestId('bar-item-3');
    fireEvent.click(fourthBarItem);

    const fourthImage = getByAltText('image-test3');
    const style = window.getComputedStyle(fourthImage);
    expect(style.opacity).toBe('1');
  });

  test('should display image 3 after press `enter` on fourth bar item', () => {
    const { getByTestId, getByAltText } = render(
      <Carousel {...defaultProps} allowAutomaticView />
    );

    const fourthBarItem = getByTestId('bar-item-3');
    fireEvent.keyDown(fourthBarItem, { key: 'Enter' });

    const fourthImage = getByAltText('image-test3');
    const style = window.getComputedStyle(fourthImage);
    expect(style.opacity).toBe('1');
  });
});
