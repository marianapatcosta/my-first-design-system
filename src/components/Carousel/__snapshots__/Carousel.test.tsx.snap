// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders without errors and matches snapshot 1`] = `
<div
  aria-label="images slider"
  class="sc-iBPRYJ bxzCfB"
  id=""
  role="listbox"
>
  <div
    class="sc-fubCfw fFQare"
  >
    <header
      class="sc-pFZIQ bIvEnm"
    >
      <h2
        class="sc-jrAGrp iyhaZy"
      >
        Image slider
      </h2>
    </header>
    <div
      class="sc-crrsfI cNTFhl"
    >
      <div
        class="sc-dQppl gHIbYh"
      >
        <div
          class="sc-dlfnbm eUChx sc-bqyKva dTXmhP"
          data-testid="image-placeholder"
        />
        <img
          alt="image-test"
          aria-label="image"
          class="sc-gsTCUz gGjrhV sc-bqyKva dTXmhP"
          data-testid="image"
          id=""
          loading="lazy"
          role="option"
          src="test.jpg"
          tabindex="-1"
        />
        <div
          class="sc-dlfnbm eUChx sc-bqyKva ijzSSL"
          data-testid="image-placeholder"
        />
        <img
          alt="image-test1"
          aria-label="image"
          class="sc-gsTCUz gGjrhV sc-bqyKva ijzSSL"
          data-testid="image"
          id=""
          loading="lazy"
          role="option"
          src="test1.jpg"
          tabindex="-1"
        />
        <div
          class="sc-dlfnbm eUChx sc-bqyKva ijzSSL"
          data-testid="image-placeholder"
        />
        <img
          alt="image-test2"
          aria-label="image"
          class="sc-gsTCUz gGjrhV sc-bqyKva ijzSSL"
          data-testid="image"
          id=""
          loading="lazy"
          role="option"
          src="test2.jpg"
          tabindex="-1"
        />
        <div
          class="sc-dlfnbm eUChx sc-bqyKva ijzSSL"
          data-testid="image-placeholder"
        />
        <img
          alt="image-test3"
          aria-label="image"
          class="sc-gsTCUz gGjrhV sc-bqyKva ijzSSL"
          data-testid="image"
          id=""
          loading="lazy"
          role="option"
          src="test3.jpg"
          tabindex="-1"
        />
        <div
          class="sc-dlfnbm eUChx sc-bqyKva ijzSSL"
          data-testid="image-placeholder"
        />
        <img
          alt="image-test4"
          aria-label="image"
          class="sc-gsTCUz gGjrhV sc-bqyKva ijzSSL"
          data-testid="image"
          id=""
          loading="lazy"
          role="option"
          src="test4.jpg"
          tabindex="-1"
        />
        <button
          aria-label="click to select previous item"
          class="sc-iqHYGH iRzAOP"
        >
          <img
            alt="left-chevron"
            src="chevron.svg"
          />
        </button>
        <button
          aria-label="click to select next item"
          class="sc-iqHYGH kloFTO"
        >
          <img
            alt="right-chevron"
            src="chevron.svg"
          />
        </button>
        <span
          class="sc-kstrdz unuVy"
        >
          <button
            aria-label="click to select item 0"
            class="sc-hBEYos jIVUEM"
            data-testid="bar-item-0"
          />
          <button
            aria-label="click to select item 1"
            class="sc-hBEYos bTMHcU"
            data-testid="bar-item-1"
          />
          <button
            aria-label="click to select item 2"
            class="sc-hBEYos bTMHcU"
            data-testid="bar-item-2"
          />
          <button
            aria-label="click to select item 3"
            class="sc-hBEYos bTMHcU"
            data-testid="bar-item-3"
          />
          <button
            aria-label="click to select item 4"
            class="sc-hBEYos bTMHcU"
            data-testid="bar-item-4"
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
